src/CustomFunction.ts

```diff
@@ -28,115 +31,78 @@ interface FunctionFailArguments {
 
 export type FunctionFailFn = (params: FunctionFailArguments) => Promise<FunctionsCompleteErrorResponse>;
 
-export interface CustomFunctionExecuteMiddlewareArgs extends SlackEventMiddlewareArgs<'function_executed'> {
+export type SlackCustomFunctionMiddlewareArgsOptions = { autoAcknowledge: boolean };
+
+export type SlackCustomFunctionMiddlewareArgs<
+  Options extends SlackCustomFunctionMiddlewareArgsOptions = { autoAcknowledge: true },
+> = SlackEventMiddlewareArgs<'function_executed', Options> & {
   inputs: FunctionExecutedEvent['inputs'];
   complete: FunctionCompleteFn;
   fail: FunctionFailFn;
-}
-
-/** Types */
-
-export type SlackCustomFunctionMiddlewareArgs = CustomFunctionExecuteMiddlewareArgs;
-
-type CustomFunctionExecuteMiddleware = Middleware<CustomFunctionExecuteMiddlewareArgs>[];
-
-export type CustomFunctionMiddleware = Middleware<CustomFunctionExecuteMiddlewareArgs>[];
-
-export type AllCustomFunctionMiddlewareArgs<
-  T extends SlackCustomFunctionMiddlewareArgs = SlackCustomFunctionMiddlewareArgs,
-> = T & AllMiddlewareArgs;
+};
 
-/** Constants */
+export function isSlackCustomFunctionMiddlewareArgsOptions<Options extends SlackEventMiddlewareArgsOptions>(
+  optionOrListener: Options | Middleware<SlackCustomFunctionMiddlewareArgs<Options>>,
+): optionOrListener is Options {
+  return typeof optionOrListener !== 'function' && 'autoAcknowledge' in optionOrListener;
+}
 
-const VALID_PAYLOAD_TYPES = new Set(['function_executed']);
+/*
+ * Middleware that filters out messages that don't match pattern
+ */
+export function matchFunction(callbackId: string): Middleware<SlackCustomFunctionMiddlewareArgs> {
+  return async ({ payload, next }) => {
+    if (payload.function.callback_id === callbackId) {
+      await next();
+    }
+  };
+}
 
 /** Class */
-
-export class CustomFunction {
+export class CustomFunction<Options extends SlackEventMiddlewareArgsOptions = { autoAcknowledge: true }> {
   /** Function callback_id */
   public callbackId: string;
 
-  private appWebClientOptions: WebClientOptions;
+  private listeners: Middleware<SlackCustomFunctionMiddlewareArgs<Options>>[];
 
-  private middleware: CustomFunctionMiddleware;
+  private options: Options;
 
-  public constructor(callbackId: string, middleware: CustomFunctionExecuteMiddleware, clientOptions: WebClientOptions) {
-    validate(callbackId, middleware);
+  public constructor(
+    callbackId: string,
+    listeners: Middleware<SlackCustomFunctionMiddlewareArgs<Options>>[],
+    options: Options,
```

Comment:
This seems like a breaking change: we change the type of the last parameter to `Options` from `WebClientOptions`. It also looks to me like we don't allow overriding the web client options anymore in this PR? Is that by design?

Reply:
I was under the impression that this `CustomFunction` class was for internal consumption only 🤔 

> It also looks to me like we don't allow overriding the web client options anymore in this PR?


Yes this is by design, the previous implementation would overwrite the `client`, this is no longer needed since the proper `token` and `client` are initialized in `App.processEvent`. This removes duplicated code 👍

Reply:
OK thanks for clarifying. Yeah I think you are right that it is for internal usage only; devs should only consume it via `app.function`.